pkg/manager/member/tikv_scaler.go

```diff
@@ -52,91 +53,187 @@ func (s *tikvScaler) Scale(meta metav1.Object, oldSet *apps.StatefulSet, newSet
 
-	if pass, err := s.preCheckUpStores(tc, podName); !pass {
-		return err
+	var (
+		upTikvStoreCount int
+		deleteStoreCount int
+		skipPreCheck     bool
+		maxReplicas      int
+	)
+	if !tc.TiKVBootStrapped() {
+		klog.Infof("TiKV of Cluster %s/%s is not bootstrapped yet, skip pre check when scale in TiKV", tc.Namespace, tc.Name)
+		skipPreCheck = true
+	} else {
+		var err error
+		pdClient := controller.GetPDClient(s.deps.PDControl, tc)
+		storesInfo, err := pdClient.GetStores()
+		if err != nil {
+			return fmt.Errorf("failed to get stores info in TidbCluster %s/%s", tc.GetNamespace(), tc.GetName())
+		}
+		config, err := pdClient.GetConfig()
+		if err != nil {
+			return fmt.Errorf("failed to get config in TidbCluster %s/%s", tc.GetNamespace(), tc.GetName())
+		}
+		maxReplicas = int(*(config.Replication.MaxReplicas))
+		// filter out TiFlash
+		for _, store := range storesInfo.Stores {
+			if store.Store != nil && store.Store.StateName == v1alpha1.TiKVStateUp && util.MatchLabelFromStoreLabels(store.Store.Labels, label.TiKVLabelVal) {
+				upTikvStoreCount++
+			}
+		}
 	}
 
-	// call PD API to delete the store of the TiKV Pod to be scaled in
-	for _, store := range tc.Status.TiKV.Stores {
-		if store.PodName == podName {
-			state := store.State
-			id, err := strconv.ParseUint(store.ID, 10, 64)
-			if err != nil {
-				return err
+	scaleInOne := func(ordinal int32) error {
+		podName := ordinalPodName(v1alpha1.TiKVMemberType, tcName, ordinal)
+		pod, err := s.deps.PodLister.Pods(ns).Get(podName)
+		if err != nil {
+			return fmt.Errorf("tikvScaler.ScaleIn: failed to get pods %s for cluster %s/%s, error: %s", podName, ns, tcName, err)
+		}
+
+		if !skipPreCheck {
+			if pass := s.preCheckUpStores(tc, podName, upTikvStoreCount, deleteStoreCount, maxReplicas); !pass {
+				return fmt.Errorf("tikvScaler.ScaleIn: failed to pass up stores check , pod %s, cluster %s/%s", podName, ns, tcName)
 			}
-			if state != v1alpha1.TiKVStateOffline {
-				if err := controller.GetPDClient(s.deps.PDControl, tc).DeleteStore(id); err != nil {
-					klog.Errorf("tikvScaler.ScaleIn: failed to delete store %d, %v", id, err)
+		}
+
+		// call PD API to delete the store of the TiKV Pod to be scaled in
+		for _, store := range tc.Status.TiKV.Stores {
+			if store.PodName == podName {
+				state := store.State
+				id, err := strconv.ParseUint(store.ID, 10, 64)
+				if err != nil {
 					return err
 				}
-				klog.Infof("tikvScaler.ScaleIn: delete store %d for tikv %s/%s successfully", id, ns, podName)
+				if state != v1alpha1.TiKVStateOffline {
+					if err := controller.GetPDClient(s.deps.PDControl, tc).DeleteStore(id); err != nil {
+						klog.Errorf("tikvScaler.ScaleIn: failed to delete store %d, %v", id, err)
+						return err
+					}
+					klog.Infof("tikvScaler.ScaleIn: delete store %d for tikv %s/%s successfully", id, ns, podName)
+					if state == v1alpha1.TiKVStateUp {
+						deleteStoreCount++
```

Comment:
How about giving a better name to indicate `TiKVStateUp`?

Reply:
any idea ?

Reply:
`deletedUpStoreCount`?